src/HttpOverStream/BodyStream.cs

```diff
@@ -75,9 +87,17 @@ public override async Task<int> ReadAsync(byte[] buffer, int offset, int count,
             }
             var read = await Underlying.ReadAsync(buffer, offset, count, cancellationToken).ConfigureAwait(false);
             _read += read;
+            CloseIfReachedEnd();
             return read;
         }
         public override void Close() => Underlying.Close();
+        protected override void Dispose(bool disposing) {

```

Comment:
This doesnt look like a normal dispose pattern, do we need the finalizer etc?

https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose

```
using System;

class BaseClass : IDisposable
{
   // Flag: Has Dispose already been called?
   bool disposed = false;
   
   // Public implementation of Dispose pattern callable by consumers.
   public void Dispose()
   { 
      Dispose(true);
      GC.SuppressFinalize(this);           
   }
   
   // Protected implementation of Dispose pattern.
   protected virtual void Dispose(bool disposing)
   {
      if (disposed)
         return; 
      
      if (disposing) {
         // Free any other managed objects here.
         //
      }
      
      // Free any unmanaged objects here.
      //
      disposed = true;
   }

   ~BaseClass()
   {
      Dispose(false);
   }
}
```

Reply:
Note that I am overriding `Stream.Dispose(bool)` there (not introducing a new base class). When calling base.Dispose(disposing). From the same article:
```
using Microsoft.Win32.SafeHandles;
using System;
using System.Runtime.InteropServices;

class DerivedClass : BaseClass
{
   // Flag: Has Dispose already been called?
   bool disposed = false;
   // Instantiate a SafeHandle instance.
   SafeHandle handle = new SafeFileHandle(IntPtr.Zero, true);

   // Protected implementation of Dispose pattern.
   protected override void Dispose(bool disposing)
   {
      if (disposed)
         return; 
      
      if (disposing) {
         handle.Dispose();
         // Free any other managed objects here.
         //
      }
      
      // Free any unmanaged objects here.
      //

      disposed = true;
      // Call base class implementation.
      base.Dispose(disposing);
   }
}
```